### PR TITLE
Add sops lookup plugin

### PR DESCRIPTION
**Pull request by mhumeSF**
##### SUMMARY
Moving PR from https://github.com/ansible/ansible/pull/59639 to here.

##### ISSUE TYPE
- Feature Pull Request=

##### COMPONENT NAME
Sops

### CODE REVIEW

File: plugins/module_utils/sops.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Copyright (c), Edoardo Tenani <e.tenani@arduino.cc>, 2018-2020
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+
+from ansible.errors import AnsibleError
```
Comment (felixfontein) [inline plugins/module_utils/sops.py:6]:
```suggestion
```

File: plugins/module_utils/sops.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    128: "CouldNotRetrieveKey",
+    111: "NoEncryptionKeyFound",
+    200: "FileHasNotBeenModified",
+    201: "NoEditorFound",
+    202: "FailedToCompareVersions",
+    203: "FileAlreadyEncrypted"
+}
+
+
+class SopsError(AnsibleError):
```
Comment (felixfontein) [inline plugins/module_utils/sops.py:41]:
```suggestion
class SopsError(Exception):
```

File: plugins/module_utils/sops.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    ''' Extend AnsibleError class with sops specific informations '''
+
+    def __init__(self, filename, exit_code, message,):
+        exception_name = sops_error_codes[exit_code]
+        message = "error with file %s: %s exited with code %d: %s" % (filename, exception_name, exit_code, message)
+        super(SopsError, self).__init__(message=message)
```
Comment (felixfontein) [inline plugins/module_utils/sops.py:47]:
```suggestion
        super(SopsError, self).__init__(message)
```

File: plugins/lookup/sops.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                        if exit_code in sops_error_codes.keys():
+                            raise SopsError(lookupfile, exit_code, err)
+                        else:
+                            raise AnsibleLookupError("could not decrypt file %s; Unknown sops error code: %s" % (to_native(term), to_native(exit_code)))
+
+                    ret.append(output.rstrip())
+                else:
+                    raise AnsibleParserError()
+            except AnsibleParserError:
+                raise AnsibleError("could not locate file in lookup: %s" % to_native(term))
```
Comment (felixfontein) [inline plugins/lookup/sops.py:105]:
```suggestion
                raise AnsibleError("could not locate file in lookup: %s" % to_native(term))
            except SopsError as e:
                raise AnsibleError(to_native(e))
```

File: plugins/module_utils/sops.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Copyright (c), Edoardo Tenani <e.tenani@arduino.cc>, 2018-2020
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
```
Comment (felixfontein) [inline plugins/module_utils/sops.py:4]:
```suggestion
from __future__ import absolute_import, division, print_function
 __metaclass__ = type
```

File: plugins/lookup/sops.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    output = to_text(output, errors='surrogate_or_strict')
+                    # the process output is the decrypted secret; be cautious
+
+                    # sops logs always to stderr, as stdout is used for
+                    # file content
+                    if err:
+                        display.vvvv(err)
+
+                    if exit_code > 0:
+                        if exit_code in sops_error_codes.keys():
```
Comment (felixfontein) [inline plugins/lookup/sops.py:96]: This `if` should go into the `SopsError` constructor.

File: plugins/lookup/sops.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                            raise SopsError(lookupfile, exit_code, err)
+                        else:
+                            raise AnsibleLookupError("could not decrypt file %s; Unknown sops error code: %s" % (to_native(term), to_native(exit_code)))
```
Comment (felixfontein) [inline plugins/lookup/sops.py:99]: In fact all these lines should go into `Sops.decrypt`.

File: plugins/module_utils/sops.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ __metaclass__ = type
```
Comment (felixfontein) [inline plugins/module_utils/sops.py:5]: I have no idea where that space came from :)
```suggestion
__metaclass__ = type
```

File: test/integration/targets/lookup_sops/files/simple.sops.yaml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+foo: ENC[AES256_GCM,data:a25L,iv:X8ILHZr+YiyLWa90Y+cwoMD1nVuel7JyTs0A5+oiOOo=,tag:GbBtp+Yqx1KEjdyztqS4EQ==,type:str]
+sops:
+    kms: []
+    gcp_kms: []
+    azure_kv: []
+    lastmodified: '2020-02-20T10:44:32Z'
+    mac: ENC[AES256_GCM,data:BAwQqD9sHgHkmlxPQLKq28Xy48qPp1B/+GDLEsIxir6WNhZgw8OgjVF1u/wCAad6qHkmN02Bwenr+aay6uKfCuOEsTRSvZ7v80yAU+h0wL3zJ/KMkRsE3QP3CWxcLQxInt+YaBjR+Q0IUjDXKm3u6ZomixZe5F5pwWr36ErV6Y0=,iv:e/iiyXQiCh8C2w/bc8mr/Psv+ehmqEMqEC1/bbGFHpY=,tag:NSDo2HISIBJhYvsqrU0mSA==,type:str]
+    pgp:
+    -   created_at: '2020-02-20T10:44:32Z'
```
Comment (felixfontein) [inline test/integration/targets/lookup_sops/files/simple.sops.yaml:9]:
```suggestion
      - created_at: '2020-02-20T10:44:32Z'
```

File: test/integration/targets/lookup_sops/tasks/ubuntu.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+- name: Test simple lookup 
+  set_fact:
+    sops_success: "{{ lookup('sops', 'simple.sops.yaml') }}"
+  ignore_errors: yes
+  register: sops_lookup_simple
+    
+- assert:
+    that:
+      - "sops_lookup_simple is success"
+
```
Comment (felixfontein) [inline test/integration/targets/lookup_sops/tasks/ubuntu.yml:34]:
```suggestion
```

File: test/integration/targets/lookup_sops/tasks/ubuntu.yml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+---
+
+- name: Test lookup with missing file
+  set_fact:
+    sops_file_does_not_exists: "{{ lookup('sops', 'file-does-not-exists.sops.yml') }}"
```
Comment (felixfontein) [inline test/integration/targets/lookup_sops/tasks/ubuntu.yml:5]:
```suggestion
    sops_file_does_not_exists: "{{ lookup('community.general.sops', 'file-does-not-exists.sops.yml') }}"
```

File: test/integration/targets/lookup_sops/tasks/ubuntu.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  register: sops_lookup_missing_file
+
+- assert:
+    that:
+      - "sops_lookup_missing_file is failed"
+      - "'could not locate file in lookup: file-does-not-exists.sops.yml' in sops_lookup_missing_file.msg"
+
+- name: Test lookup of non-sops file
+  set_fact:
+    sops_wrong_file: "{{ lookup('sops', 'wrong.yaml') }}"
```
Comment (felixfontein) [inline test/integration/targets/lookup_sops/tasks/ubuntu.yml:16]:
```suggestion
    sops_wrong_file: "{{ lookup('community.general.sops', 'wrong.yaml') }}"
```

File: test/integration/targets/lookup_sops/tasks/ubuntu.yml [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+  register: sops_lookup_wrong_file
+
+- assert:
+    that:
+      - "sops_lookup_wrong_file is failed"
+      - "'sops metadata not found' in sops_lookup_wrong_file.msg"
+
+- name: Test simple lookup
+  set_fact:
+    sops_success: "{{ lookup('sops', 'simple.sops.yaml') }}"
```
Comment (felixfontein) [inline test/integration/targets/lookup_sops/tasks/ubuntu.yml:27]:
```suggestion
    sops_success: "{{ lookup('community.general.sops', 'simple.sops.yaml') }}"
```

File: plugins/lookup/sops.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError, AnsibleParserError, AnsibleLookupError
+from ansible.plugins.lookup import LookupBase
+from ansible.module_utils._text import to_text, to_native
+from ansible.module_utils.sops import Sops, SopsError, sops_error_codes
```
Comment (felixfontein) [inline plugins/lookup/sops.py:27]:
```suggestion
from ansible_collections.community.general.plugins.module_utils.sops import Sops, SopsError, sops_error_codes
```

File: plugins/lookup/sops.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError, AnsibleParserError, AnsibleLookupError
```
Comment (felixfontein) [inline plugins/lookup/sops.py:24]:
```suggestion
from ansible.errors import AnsibleLookupError
```

File: plugins/lookup/sops.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+                        else:
+                            raise AnsibleLookupError("could not decrypt file %s; Unknown sops error code: %s" % (to_native(term), to_native(exit_code)))
+
+                    ret.append(output.rstrip())
+                else:
+                    raise AnsibleParserError()
+            except AnsibleParserError:
+                raise AnsibleError("could not locate file in lookup: %s" % to_native(term))
+            except SopsError as e:
+                raise AnsibleError(to_native(e))
```
Comment (felixfontein) [inline plugins/lookup/sops.py:107]:
```suggestion
                raise AnsibleLookupError(to_native(e))
```

File: plugins/lookup/sops.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+                        if exit_code in sops_error_codes.keys():
+                            raise SopsError(lookupfile, exit_code, err)
+                        else:
+                            raise AnsibleLookupError("could not decrypt file %s; Unknown sops error code: %s" % (to_native(term), to_native(exit_code)))
+
+                    ret.append(output.rstrip())
+                else:
+                    raise AnsibleParserError()
+            except AnsibleParserError:
+                raise AnsibleError("could not locate file in lookup: %s" % to_native(term))
```
Comment (felixfontein) [inline plugins/lookup/sops.py:105]:
```suggestion
                    raise AnsibleLookupError("could not locate file in lookup: %s" % to_native(term))
```

**Comment by mhumeSF** [comment]
Whoops... messed up rebase. Thanks for catching `test/tests` one sec.

**Comment by felixfontein** [comment]
The integration tests under Ubuntu 16.04 are failling: https://app.shippable.com/github/ansible-collections/community.general/runs/1584/25/tests
For Ubuntu 18.04 they are passing though.

**Comment by felixfontein** [comment]
CI is now passing (at least everything that's related to this PR - the failures are unrelated). Cool!

**Comment by felixfontein** [comment]
If nobody asks for changes until the weekend, I'll merge this.

**Comment by felixfontein** [comment]
recheck

**Comment by felixfontein** [comment]
Tests pass. Merging.

**Comment by felixfontein** [comment]
@endorama thanks a lot for contributing this lookup plugin!
@mhumeSF thanks a lot for porting the PR to community.general and helping getting it merged!

**Comment by endorama** [comment]
@felixfontein :bow: thank you for the continuous effort on this and guiding me through it!
@mhumeSF thank you too for the effort!

**Pull request merged**